native,cgo: add Dgeev

vladimir-ch opened the pull request:
I still want to extend the testing a bit but you can start looking at this, especially the testing.

btracey commented:
Wooooooo!

btracey commented:
This looks like a real failure.

btracey commented on native/dgeev.go:
```diff
@@ -0,0 +1,283 @@
+		impl.Dorghr(n, ilo, ihi, vl, ldvl, tau, work[iwrk:], lwork-iwrk)
+		// Perform QR iteration, accumulating Schur vectors in VL.
+		iwrk = n
+		firstok = impl.Dhseqr(lapack.EigenvaluesAndSchur, lapack.UpdateZ, n, ilo, ihi,
```
Make this firstOk or firstOK? I was reading it as "first tok" for most of the code and was really confused (yes, I realize that is nonsensical)

kortschak replied:
`validFrom` or even just `first`?

vladimir-ch replied:
Went with `first`. Done.

btracey commented on native/dgeev.go:
```diff
@@ -0,0 +1,283 @@
+
+	if wantvl || wantvr {
+		// Compute left and/or right eigenvectors.
+		impl.Dtrevc3(side, lapack.AllEigVecMulQ, nil, n,
```
What does the fortran keyword do here?

kortschak replied:
`SELECT`? It's not a keyword here, it's a []bool equivalent.

kortschak commented on lapack.go:
```diff
@@ -156,3 +156,15 @@ const (
+
+// Job types for Dgeev.
+type (
+	JobLeftEV  byte
```
What's the rationale for these additions?

vladimir-ch replied:
Both jobs take the same values ('V' and 'N'), so if they are just the generic `lapack.Job`, I think it would be too easy for the caller to get them in the wrong order without any compiler error or runtime panic.

kortschak replied:
OK

kortschak commented on native/dgeev.go:
```diff
@@ -0,0 +1,283 @@
+		impl.Dhseqr(lapack.EigenvaluesAndSchur, lapack.UpdateZ, n, 0, n-1,
+			nil, 1, nil, nil, nil, 1, work, -1)
+		maxwrk = max(maxwrk, max(n+1, n+int(work[0])))
+		if wantvl {
```
```
job := lapack.RightEigVec
if wantvl {
    job = lapack.LeftEigVec
}
impl.Dtrevc3(job, lapack.AllEigVecMulQ, nil, n,
    nil, 1, nil, 1, nil, 1, n, work, -1)
```

vladimir-ch replied:
Done.

kortschak commented on native/dgeev.go:
```diff
@@ -0,0 +1,283 @@
+
+	// Scale A if max element outside range [smlnum,bignum].
+	anrm := impl.Dlange(lapack.MaxAbs, n, n, a, lda, nil)
+	scalea := false
```
`var scalea bool`?

vladimir-ch replied:
Done.

btracey commented on testlapack/dgeev.go:
```diff
@@ -0,0 +1,693 @@
+			return
+		}
+		used := make([]bool, n)
+		for i := first; i < n; i++ {
```
Does this test actually work? If any of the eigenvalues match, idx != -1, so the error check below won't work as far as I can tell. Seems like the test should be about if all of the values of `used` are set. It's also weird that `i` indexes from first, but evWant indexes from 0.

vladimir-ch replied:
The idea is just to take each computed eigenvalue and see if there is an unused known eigenvalue sufficiently close to it. If there is not (i.e., idx == -1), then error. The order of the eigenvalues is unknown and if first > 0, then it is not known which eigenvalues were actually computed, so I can't think of anything better than this. This is also why `i` indexes from first (the index of the first valid eigenvalue), which evWant from 0 (the computed eigenvalue can be any element of `test.evWant`).

Of course, this test assumes that the known eigenvalues are well separated.

btracey replied:
I see, okay

btracey commented on testlapack/dgeev.go:
```diff
@@ -0,0 +1,693 @@
+		return
+	}
+
+	for k := 0; k < n; {
```
Add comment.

vladimir-ch replied:
Done.

vladimir-ch commented on cgo/lapack.go:
```diff
@@ -1712,3 +1716,111 @@ func (impl Implementation) Dhseqr(job lapack.Job, compz lapack.Comp, n, ilo, ihi
+// The right eigenvector v_j of A corresponding to an eigenvalue λ_j
+// is defined by
+//  A v_j = λ_j v_j,
+// and the left eigenvector u_j is defined by
```
Done.

vladimir-ch commented on native/dgeev.go:
```diff
@@ -0,0 +1,283 @@
+// same order as their eigenvalues. If the j-th eigenvalue is real, then
+//  u_j = VL[:,j],
+//  v_j = VR[:,j],
+// and if the j-th and (j+1)-th eigenvalues form a complex conjugate pair, then
```
Done.

vladimir-ch commented on testlapack/dgeev.go:
```diff
@@ -0,0 +1,693 @@
+		}
+	}
+
+	// for _, n := range []int{10, 30, 50} {
```
I expected that commenting this out would trigger questions which in turn would provide a place for explanation and discussion.

It depends. The test matrices above are nice but many of them have too simple structure that will not trigger the inner parts of the algorithm. I wanted to have a couple of tests on (random) matrices with known eigenvalues that are a mix of real and complex conjugate pair numbers. Starting with the real Schur decomposition theorem that for every nxn matrix there exists an orthogonal matrix Q such that 'Q^T A Q = T' where T is in real Schur form (quasi upper triangular with 1x1 and 2x2 blocks on the diagonal), I generate a random T and a random Q and multiply them to get my random A with known eigenvalues. It works but the computed eigenvalues are rather inaccurate, I would have to use tolerance 1e-4 or even 1e-2 (it gets worse with increasing n and affects both real and complex eigenvalues). That's suspiciously loose and that's why I commented it out.

Possible places to look at:
- I'm not using the real Schur decomposition theorem (for every matrix there exists ...), rather "for any T in real Schur form and Q orthogonal, `Q T Q^T` has the same eigenvalues as T". I guess it's fine but I don't have a proof that it holds. I have seen using this approach to generate random symmetric matrices with given eigenvalues, but not for general non-symmetric matrices.
- There might be something wrong with `reflector` and `applyReflector` functions although it does seem that the matrices are indeed orthogonal.
- The randomness creates bad matrices so that `Q T Q^T` has slightly different eigenvalues than T. I suspect this is the reason.

I'll try to think about it further but any comments are appreciated anyway.

btracey replied:
Why don't you generate random eigenvalues and eigenvectors, multiply the matrices, and then run the algorithm?

vladimir-ch replied:
I've tried using the Jordan canonical form with the same result. I've also tried replacing `reflector` and `applyReflector` with generating a random orthogonal matrix and passing it through `Dgeqrf` and `Dorgqr` with the same result. Also, `cgo` behaves identically. All this kind of confirms my initial suspicion that `Q T Q^T` simply has slightly different eigenvalues from those of `T`  due to the matrix multiplication . Without known eigenvalues this test becomes worthless so I suggest removing it. @btracey ?

vladimir-ch commented on testlapack/dgeev.go:
```diff
@@ -0,0 +1,693 @@
+	wr := make([]float64, n)
+	wi := make([]float64, n)
+
+	lwork := max(1, 3*n)
```
I'll try to do something about it.

vladimir-ch replied:
Ok, done. PTAL @btracey

vladimir-ch commented:
I think this is getting stabilized. Coverage is almost complete (I still can't get any non-converged eigenvalue), I've added testing with an intermediate workspace size, and I've added Geev to `lapack64`. PTAL

btracey commented:
LGTM, but wait for @kortschak

kortschak commented:
LGTM with a minimal at the tests.

As an aside, the naming of the implementation type is a little odd. Presumably eventually we will autogenerate a float32 implementation as well. What is the plan for naming when that happens?

vladimir-ch commented:
What implementation type do you mean?

kortschak commented:
Sorry. Brain fart.
